### PR TITLE
ccgx: Retry the HID disconnect up to 5 times

### PR DESCRIPTION
**Pull request by hughsie**
Based on a patch by  Ryan Lee <ryle@cypress.com>, many thanks.

### CODE REVIEW

File: plugins/ccgx/fu-ccgx-hpi-device.c [hunk 1]
```diff
@@ -303,46 +296,51 @@ fu_ccgx_hpi_device_i2c_read (FuCcgxHpiDevice *self,
...
+		.mode = CY_I2C_MODE_READ,
+		.buf = buf,
+		.bufsz = bufsz,
+		.cfg_bits = cfg_bits,
+	};
+	return fu_device_retry (FU_DEVICE (self),
+				fu_ccgx_hpi_device_i2c_read_cb,
+				HPI_CMD_RESET_RETRY_CNT,
+				&helper, error);
+}
```
[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/ccgx/fu-ccgx-hpi-device.c:314]: My mistake, apologies. Will fix now.

File: plugins/ccgx/fu-ccgx-hpi-device.c [hunk 2]
```diff
@@ -234,14 +237,14 @@ fu_ccgx_hpi_device_wait_for_notify (FuCcgxHpiDevice *self,
...
 				     "i2c status error in i2c write [0x%x] event: %s",
 				     (guint8) buf[0], error_local->message);
 		/* read */
 		} else {
 			g_set_error (error,
 				     FWUPD_ERROR,
-				     FWUPD_ERROR_INTERNAL,
+				     FWUPD_ERROR_READ,
 				     "i2c status error in i2c read [0x%x] event: %s",
 				     (guint8) buf[0], error_local->message);
```
[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/ccgx/fu-ccgx-hpi-device.c:249]: Fixed, thanks.

File: plugins/ccgx/fu-ccgx-hpi-device.c [hunk 3]
```diff
@@ -68,8 +70,10 @@ fu_ccgx_hpi_device_to_string (FuDevice *device, guint idt, GString *str)
...
 typedef struct {
-	guint8		 mode;
-	guint8		*i2c_status;
+	guint8			 mode;
+	guint8			*buf;
+	gsize			 bufsz;
+	CyI2CDataConfigBits	 cfg_bits;
 } FuCcgxHpiDeviceRetryHelper;
 
 static gboolean
```
[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/ccgx/fu-ccgx-hpi-device.c:79]: Fixed, thanks.

File: plugins/ccgx/fu-ccgx-hpi-device.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/ccgx/fu-ccgx-hpi-device.c]: @chlee75 something like this?

**Comment by hughsie** [comment]
Don't merge this yet; there's a regression.

**Comment by hughsie** [comment]
> Don't merge this yet; there's a regression.

The device regression test gets its first kill: fixed. :)

**Comment by chlee75** [comment]
Do not need g_usleep (HPI_CMD_FLASH_READ_WRITE_DELAY_US) in fu_ccgx_hpi_device_reg_read_cb(), fu_ccgx_hpi_device_reg_write_cb(), fu_ccgx_hpi_device_reg_write_no_resp.

30 msec is too big in every i2c read and write operation.
5 or 10 msec would be OK.

But about 30 msec delay is needed in fu_ccgx_hpi_device_i2c_reset_cb().

**Comment by hughsie** [comment]
> But about 30 msec delay is needed in fu_ccgx_hpi_device_i2c_reset_cb().

This is covered by `fu_device_retry_set_delay (FU_DEVICE (self), HPI_CMD_RETRY_DELAY);`

> 30 msec is too big in every i2c read and write operation.

What about the last patch on this PR -- better?

**Comment by chlee75** [comment]
> 30 msec is too big in every i2c read and write operation.
> 
> What about the last patch on this PR -- better?

It looks good to me. 
one thing, could you replace HPI_CMD_FLASH_READ_WRITE_DELAY_US with HPI_CMD_REG_READ_WRITE_DELAY_US?
The delay actually is for register read or write not for flash read or write.

**Comment by hughsie** [comment]
> could you replace HPI_CMD_FLASH_READ_WRITE_DELAY_US with HPI_CMD_REG_READ_WRITE_DELAY_US

Done!

**Pull request merged**